lrlex/src/lib/lexer.rs

```diff
@@ -651,4 +687,115 @@ b 'B'
             .collect::<Vec<_>>();
         assert_eq!(anonymous_rules[0].name_span, Span::new(21, 21));
     }
+
+    #[test]
+    fn test_token_start_states() {
+        let src = "%x EXCLUSIVE_START
```

Comment:
What does `%exclusive_start` mean vs. `%inclusive_start`?

Reply:
This is perhaps the feature I miss the most, so thank you for working on it Simon!
The flex documentation is here:
https://westes.github.io/flex/manual/Start-Conditions.html#Start-Conditions

The short answer is that the inclusive state `<Inc>` state matches *both* `b` and  `c`, while the exclusive only matches `a`
```
%x Ex
%s Inc
<Ex>a "A"
<Inc>b "B"
c "C"
```

It is really useful for lexing things like SML style nested comments `(* (* commented out comment *) outer comment *)`
and probably rust-like raw string literals `r#"""#`, `r##""""##`.

I haven't used it with a lexer that doesn't have actions, e.g. in flex you can see that switching between states is generally done in action code.  I haven't thought too much about it, but given the actionless style of lrlex, in both the cases I mentioned it seems there is a counter number of `#` signs and the number of open comments.

E.g. we would re-enter the initial state after seeing the appropriate number of '#'s preceeded by a quote, or a balanced number of `(*` and `*)`.

I guess I struggle somewhat to see transformations for doing this in an actionless way.  Towards the end of that flex documentation it mentions `yy_push_state`/`yy_pop_state`.

In the calc example in your commit, there is these snippets below switching states,
```
/\*            <COMMENT>;
<COMMENT>\*/   <INITIAL>;
```

I would be kind of curious about a push/pop like version of these, with `+` for push and `-` for pop.
```
/\*            +<COMMENT>;
<COMMENT>\*/   -<COMMENT>;
```

And whether something like that would allow for a translation of these generally counter dependent state exits.

Reply:
Ah! I have never used these. AIUI these are part of "normal" lex and aren't a flex extension? I'm certainly happy to improve compatibility with lex!

Reply:
I'm uncertain about that, some reading (o'reilly lex & yacc) seems to indicate that at least `exclusive` states originate with *flex*, and insinuates that some have `inclusive` without `exclusive`.  The *POSIX* lex specification doesn't seem to describe either `inclusive` or `exclusive` as far as I can tell.

Reply:
Right. I've long thought (and there was a partial PR from someone to do this) that lex should, like lrpar, support different flavours of lexer. This might be a good candidate for that. There again, lexing is simpler than parsing, and given we don't formally support lex's regex language, pretending that we have greater compatibility than we actually have might be a fool's mission!

Reply:
Still some code clean-up work I want to do on this work (plus rebasing onto most recent commits in trunk).

Reading https://pubs.opengroup.org/onlinepubs/007904875/utilities/lex.html, it mentions both inclusive and exclusive start states (or start conditions, they appear to be used interchangeably).

Would it be worth me adding some documentation into this Pull Request - how inclusive and exclusive start states work, with an example?

Reply:
Pointing to the definitions elsewhere of inclusive/exclusive is fine with me! We will, of course, need tests that cover all the possibilities so that any code proposed for merging does what we think it should :)

[I really need to duplicate this for grmtools, but while I'm here: the process we use for PRs is documented at https://ykjit.github.io/ykdocs/contributing/prs.html. Please let me know if things aren't clear!]

Reply:
I believe this is ready for a review - I've added the tests I consider necessary, and updated the documentation.
I've been consulting the flex source (especially https://github.com/westes/flex/blob/master/src/scan.l), and it doesn't explicitly include **%start** for declaring start conditions (in the definitions section) - should I adjust to include the possibilities as in http://dinosaur.compilertools.net/lex/index.html (Section 10)?

Reply:
> should I adjust to include the possibilities as in http://dinosaur.compilertools.net/lex/index.html (Section 10)?

Oops, sorry, I missed this question! I don't think a PR needs to do everything in one go: making things better is enough :) I would, though, like to make sure that we don't do anything that's in conflict with the "standard" -- but I'm happy for you to say whether you think the PR is consistent with the spec or not.